### ISSUE TITLE
Introduce SamplingDatasetOp to TensorFlow

### ISSUE THREAD

**Issue by delock** [tensorflow/tensorflow#23179]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (1.10):
- Are you willing to contribute it (Yes):

**Describe the feature and the current behavior/state.**
I propose to introduce a SamplingDatasetOp to TensorFlow
**Will this change the current api? How?**
Yes.
User would be able to create sampling dataset that output a subset of input dataset that is uniformed sampled from input dataset, with sample rate supplied as parameter.
`dataset = dataset.sampling(rate)`
**Who will benefit with this feature?**
Minigo training have very slow filling rate. https://github.com/tensorflow/minigo/issues/520
This is due to Minigo use a FilterDatasetOp to filter out subset of input data, which has low efficiency.  With sampling dataset, we can implement the sample action inside the sampling dataset op, without call predicate and go through lambda function explicitly, which improves training performance by 1.7x.  More details can be seen in the issue for minigo above.
**Any Other info.**

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by mrry** [comment]
Hmm, it seems like some of the recent static optimizations for map/filter fusion and short-circuit maps should be able to take care of this without adding a new transformation.

Can you please post a minimal reproduction, so we can test the optimizations?

Assigning this to @jsimsa, since he added the short-circuit map support and can point you to the right optimizations.

**Comment by delock** [comment]
Here is a code snipet that reproduce this problem.
```
import tensorflow as tf                                                     
dataset = tf.data.Dataset.range(10000)                                      
dataset = dataset.filter(lambda x: tf.less(tf.random_uniform([1]), 0.05)[0])
dataset = dataset.repeat()                                                  
dataset = dataset.shuffle(buffer_size=100000)                               
                                                                            
with tf.Session() as sess:                                                  
    value = dataset.make_one_shot_iterator().get_next()                     
    sess.run(value)                                                         

```
The time output of this program:
```
~/test/tf_dataset$ time python3 tf_dataset.py 
2018-10-25 15:01:57.602520: I tensorflow/core/common_runtime/process_util.cc:69] Creating new thread pool with default inter op setting: 2. Tune using inter_op_parallelism_threads for best performance.
2018-10-25 15:02:07.623575: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 22660 of 100000
2018-10-25 15:02:17.623406: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 45551 of 100000
2018-10-25 15:02:27.624088: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 68033 of 100000
2018-10-25 15:02:37.624411: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 90821 of 100000
2018-10-25 15:02:41.167028: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:135] Shuffle buffer filled.

real    0m45.482s
user    1m18.749s
sys     0m15.244s

```

**Comment by delock** [comment]
A more complete test case that use the sampling dataset I implemented locally.  I kinda of magnify the problem by set rate to 0.01, instead of 0.05
```
import tensorflow as tf                                                           
import time                                                                       
                                                                                  
rate = 0.01                                                                       
                                                                                  
datasetS = tf.data.Dataset.range(10000)                                           
datasetS = datasetS.repeat()                                                      
datasetS = datasetS.sampling(rate)                                                
datasetS = datasetS.shuffle(buffer_size=100000)                                   
                                                                                  
datasetF = tf.data.Dataset.range(10000)                                           
datasetF = datasetF.repeat()                                                      
datasetF = datasetF.filter(lambda x: tf.less(tf.random_uniform([1]), rate)[0])    
datasetF = datasetF.shuffle(buffer_size=100000)                                   
                                                                                  
with tf.Session() as sess:                                                        
    valueS = datasetS.make_one_shot_iterator().get_next()                         
    valueF = datasetF.make_one_shot_iterator().get_next()                         
    print ("run with sampling dataset")                                           
    start = time.time()                                                           
    sess.run(valueS)                                                              
    end = time.time()                                                             
    print ("%f seconds"%(end-start))                                              
    start = time.time()                                                           
    print ("run with filter dataset")                                             
    sess.run(valueF)                                                              
    end = time.time()                                                             
    print ("%f seconds"%(end-start))                                              

```
The output shows sampling dataset op would be 47x faster than use a filter dataset op to do the job.
```
2018-10-25 15:20:41.554439: I tensorflow/core/common_runtime/process_util.cc:69] Creating new thread pool with default inter op setting: 2. Tune using inter_op_parallelism_threads for best performance.        
run with sampling dataset
3.914848 seconds
run with filter dataset
2018-10-25 15:20:55.497627: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 5451 of 100000                                                           
2018-10-25 15:21:05.498753: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 10959 of 100000                                                          
2018-10-25 15:21:15.498401: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 16433 of 100000                                                          
2018-10-25 15:21:25.498502: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 21692 of 100000                                                          
2018-10-25 15:21:35.497819: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 27089 of 100000                                                          
2018-10-25 15:21:45.500435: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 32521 of 100000                                                          
2018-10-25 15:21:55.500459: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 37949 of 100000                                                          
2018-10-25 15:22:05.499758: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 43421 of 100000                                                          
2018-10-25 15:22:15.497833: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 48904 of 100000                                                          
2018-10-25 15:22:25.508743: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 54146 of 100000                                                          
2018-10-25 15:22:35.500528: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 59416 of 100000                                                          
2018-10-25 15:22:45.498792: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 64705 of 100000                                                          
2018-10-25 15:22:55.500804: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 69902 of 100000                                                          
2018-10-25 15:23:05.498479: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 75305 of 100000                                                          
2018-10-25 15:23:15.498075: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 80657 of 100000                                                          
2018-10-25 15:23:25.499887: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 86016 of 100000                                                          
2018-10-25 15:23:35.500111: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 91404 of 100000                                                          
2018-10-25 15:23:45.499380: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 96791 of 100000                                                          
2018-10-25 15:23:51.490890: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:135] Shuffle buffer filled.
186.004482 seconds

```

**Comment by jsimsa** [comment]
Thank you for providing the example.

When I run the filter-based pipeline on my workstation, it takes roughly 160 seconds.

IIUC, the equivalent sampling logic can be implemented using the following pipeline:

```
  ds = tf.data.Dataset.range(10000).shuffle(10000)
  ds = ds.repeat()
  ds = ds.window(100).flat_map(lambda x: x.take(1))
  ds = ds.shuffle(buffer_size=100000)
```

Running this pipeline on my machine takes less than 10 seconds, so I am leaning towards not introducing new API for something that can be implemented within a factor of 2 using the existing API.

**Comment by jsimsa** [comment]
Closing this issue, assuming it is resolved, but feel free to re-open if you have follow up.

**Comment by delock** [comment]
Hi, the first step (Dataset.range(10000)) is an artificial op to expose the performance bottleneck.  In real life scenario (https://github.com/mlperf/training/blob/master/reinforcement/tensorflow/minigo/preprocessing.py#L139), data is read from file and number of elements are not predefined.  It looks to me the method @jsimsa only works for fixed length dataset, and not working for variable length dataset or dataset that read from a stream.   How do I do something methematically equivalent or close when input data size is unknown?

**Comment by delock** [comment]
I don't have access to reopen.  Do you have suggestion for my scenario?

**Comment by jsimsa** [comment]
I think the general pattern would be:

```
dataset # this is the dataset you wish to sample with ratio a/b
dataset = dataset.shuffle(x) # x trades off mathematical accuracy and performance
dataset = dataset.repeat()
dataset = dataset.window(b).flat_map(lambda x: x.take(a))
```

Does that make sense?

**Comment by delock** [comment]
I think the hard part is find out 'x' that balance between mathematical accuracy and performance.  i.e.  In the lab log of minigo, I read 'move shuffle buffer to 1M from 200k', then some later logs to move it back to 200k again, etc.  That I saw struggling between performance and accuracy.  While sampling op is both mathematical accurate and fast, which is the whole point.

**Comment by jsimsa** [comment]
I am opposed to adding `sample` to the public tf.data API. It does not add any new functionality.

As far as I can tell, you want to introduce the transformation because of performance. If that's the case, let me suggest an alternative route. You implement the C++ `SampleDataset` kernel, but instead of surfacing it in the public API, you also implement  a static optimization that replaces instances of `FilterDataset` with a "sampling" predicate (e.g. `lambda x: tf.less(tf.random_uniform([1]), rate)[0])`) with `SampleDataset`. This way, the optimization happens automatically and we do not clutter the API with endpoints that do not provide new functionality.

For an example of tf.data static optimization see https://github.com/tensorflow/tensorflow/blob/00ae12ad8bf5c348e4c31448e3922cbaab54cc03/tensorflow/core/grappler/optimizers/data/map_and_batch_fusion.cc (this optimization replaces `map(...).batch(...)` with `map_and_batch(...)`, also for the sake of performance).

**Comment by jsimsa** [comment]
`skip` and `take` are separate API because they can only be implemented using a *stateful* function and stateful functions have many limitations (e.g. do not support checkpointing and preclude many static optimizations).

If you implement my suggestion, I will be happy to review the PR.

**Comment by delock** [comment]
Hi @jsimsa 
I'm optimizing the pattern as you suggested.  However I met one problem. The `tf.random_uniform` op itself is stateful, which will be blocked by `tensorflow/core/framework/dataset.cc:AddFunction`.

It is tricky here, from definition in TensorFlow, random_uniform is stateful because two call with same input will generate different output, which is the nature of _random_.   The pseudo random number generator also requires a internal state to generate random number sequence.  On the other hand, if `random_uniform` is implemented to imitate true random number generator, each time we call `random_uniform`, we expect the output to be independent to previous outputs, make it actually stateless in another sense.  The final result is in practice this pattern cannot be statically optimized, but in spirit, this pattern should be able to be statically optimized (replaced by a `SamplingDataset`).

What's your suggestion here?  It looks like either we need to remove some restriction on optimization in TensorFlow, or we have to introduce new API for same performance purpose.

**Comment by jsimsa** [comment]
I am not sure what do you mean "will be blocked by ...". When we use `AsGraphDefInternal` to build the graph for the sake of optimization, then stateful actions are allowed (see [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/dataset.cc#L143)).

Your optimization should check whether the FilterDataset function matches a certain pattern and if so, replace it with SamplingDataset.

**Comment by delock** [comment]
Oh, I'm experimenting with r1.10.1 branch, the allow optimization part is not present on r1.10.1.  I'll try the line in master and see how it goes.
> I am not sure what do you mean "will be blocked by ...". When we use `AddFunction` to build the graph for the sake of optimization, then stateful actions are allowed (see [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/dataset.cc#L143)).
> 
> Your optimization should check whether the FilterDataset function matches a certain pattern and if so, replace it with SamplingDataset.
Yes, that's exactly I'm doing, it works well for the filter function, and get the performance expected, thanks for the suggestion!

**Comment by delock** [comment]
> `skip` and `take` are separate API because they can only be implemented using a _stateful_ function and stateful functions have many limitations (e.g. do not support checkpointing and preclude many static optimizations).
> 
> If you implement my suggestion, I will be happy to review the PR.
Hi @jsimsa , PR 25384 is opened following your suggestion.  Can you review this PR?  Thanks!
https://github.com/tensorflow/tensorflow/pull/25384
